### PR TITLE
ENH: Add headers paramater to read_json and read_csv

### LINKED ISSUE #36688: ENH: Change Pandas User-Agent and add possibility to set custom http_headers to pd.read_* functions
Currently Pandas makes HTTP requests using "Python-urllib/3.8" as a User Agent.
This prevents from downloading some resources and static files from various places.
What if, Pandas would make requests using "Pandas/1.1.0" headers instead?
There should be possibility to add custom headers too (`auth`, `csrf tokens`, `api versions` and so on).

Use Case:

I am writing a book on Pandas:
- https://python.astrotech.io/numerical-analysis/index.html#pandas

I published data in CSV and JSON to use in code listings:
- https://python.astrotech.io/_static/iris.json
- https://python.astrotech.io/_static/iris.csv

You can access those resources via browser, `curl`, or even `requests`, but not using Pandas.
The only change you'd need to do is to set User-Agent.
This is due to the `readthedocs.io` blocking "Python-urllib/3.8" User Agent for whatever reason.
The same problem affects many other places where you can get data (not only `readthedocs.io`).

Currently I get those resources with `requests` and then put `response.text` to one of:
- `pd.read_csv`
- `pd.read_json`
- `pd.read_html`

Unfortunately this makes even simplest code listings... quite complex (due to the explanation of `requests` library and why I do this like that).

Pandas uses `urllib.request.urlopen` which does not allow to set `http_headers`
https://github.com/pandas-dev/pandas/blob/master/pandas/io/common.py#L146

Although `urllib.request.urlopen` can take `urllib.request.Request` as an argument.
And `urllib.request.Request` object has possibility to set custom `http_headers`
https://docs.python.org/3/library/urllib.request.html#urllib.request.Request

Possibility to add custom `http_headers` should be in `pd.read_csv`, `pd.read_json` and `pd.read_html` functions.

From what I see, the `read_*` call stack is three to four function deep.
There are only 6 references in 4 files to `urlopen(*args, **kwargs)` function.
So the change shouldn't be quite hard to implement.

`http_headers` parameter can be `Optional[List]` which will be fully backward compatible and would not require any changes to others code.

### PR DESCRIPTION
**Pull request by Antetokounpo**
- [x] closes #36688
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This adds the option to specify headers when reading a csv or a json file from an URL in Python3.
Let me know if new tests are needed.

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -150,7 +150,14 @@ def urlopen(*args, **kwargs):
     """
     import urllib.request
 
-    return urllib.request.urlopen(*args, **kwargs)
+    # Request class is only available in Python3, which
+    # allows headers to be specified
+    if hasattr(urllib.request, "Request"):
```
Comment (rhshadrach) [inline pandas/io/common.py:155]: pandas no longer supports Python2

File: pandas/io/common.py [hunk 2]
```diff
@@ -176,6 +183,7 @@ def get_filepath_or_buffer(
     compression: CompressionOptions = None,
     mode: ModeVar = None,  # type: ignore[assignment]
     storage_options: StorageOptions = None,
+    headers: dict = {},
```
Comment (rhshadrach) [inline pandas/io/common.py:186]: Can we be more specific using Dict? What types are the keys and values?

Comment (Antetokounpo) [reply pandas/io/common.py:186]: It's HTTP Headers, so it should be `str` for keys, and the type for value can vary I think.

File: pandas/io/common.py (hunk 1 above)
Comment (Antetokounpo) [reply pandas/io/common.py:155]: I'll remove the check then.

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you add tests for the change? That is typically the first thing we look for as reviewers

Also not sure about adding this to the API - what values would a user typically provide here? Documenting that in the docstrings would help for sure

File: pandas/io/common.py [hunk 3]
```diff
@@ -148,9 +148,9 @@ def urlopen(*args, **kwargs):
     Lazy-import wrapper for stdlib urlopen, as that imports a big chunk of
     the stdlib.
     """
-    import urllib.request
+    from urllib.request import Request, urlopen as _urlopen
```
[the remark this answers is not in the document]
Comment (Antetokounpo) [reply pandas/io/common.py:151]: I find it makes the code more concise, but if you think the old way is better I can change it back.

**Comment by WillAyd** [comment]
I'm not sure we would add new arguments to the function signatures like this. There has been some updates to the original issue and I think would prefer to just document how this can be handled instead of changing these functions - can you adjust the PR accordingly?

**Comment by arw2019** [comment]
Closing in favor of #37966

**Pull request closed without merging**